# Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# Import Data

In [2]:
# setting a path
path = '/Users/tannu/Desktop/Data Analytics/Instacart Basket Analysis Dec-2025 '

In [3]:
# importing data
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '2 Data', 'Prepared Data', 'ords_prods_merge_new.pkl'))

## Q. Aggregate the mean of the “order_number” column grouped by “department_id” for your entire dataframe. 

In [4]:
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### A. The overall trend remains the same for the subset and the entire dataframe. The mean values corresponding to each department_id in both cases have a +/- 1 difference in most cases except for department_id 21. In the subset, the corresponding mean value is 25.535479, and in the full datatframe it is 22.902379. But the mean value for department_id 21 is still the highest among all departments.  

## Q. Create a loyalty flag for existing customers using the transform() and loc() functions.

In [5]:
# creating a new column 'max_order' which will display the maximum number of orders placed by each user. 
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/z8/12vdbzn57fx_ky9_g3vm3lv80000gn/T/ipykernel_25838/1709179830.py:2: FutureWarning: The provided callable <function max at 0x107541800> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [6]:
# checking if the column is created
df_ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_the_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range', 'busiest_day', 'busiest_hours',
       'max_order'],
      dtype='object')

In [7]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   3139998      138            28                       6                 11   
1   1977647      138            30                       6                 17   
2    389851      709             2                       0                 21   
3    652770      764             1                       3                 13   
4   1813452      764             3                       4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

         price_range     busiest_day   busiest_hours  max_order  
0  Mid-range product  Regularly Busy     Most Orders         32  
1  Mid-range product  Regularly Busy  Average Orders         32  
2  Mid-range product     Busiest Day  Average Orders          5  
3  Mid-range product  Regularly Busy     Most Orders          3  
4  Mid-range product      Least Busy  Average Orders          3

In [8]:
# creating a new column 'loyalty_flag' to flag customer loyalty by using the values in 'max_order'
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

In [9]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular Customer'

In [10]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New Customer'

In [12]:
# Checking if 'loyalty_flag' column is created
df_ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_the_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range', 'busiest_day', 'busiest_hours',
       'max_order', 'loyalty_flag'],
      dtype='object')

In [11]:
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: count, dtype: int64

In [14]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_the_week  order_hour_of_day  \
0   3139998      138            28                       6                 11   
1   1977647      138            30                       6                 17   
2    389851      709             2                       0                 21   
3    652770      764             1                       3                 13   
4   1813452      764             3                       4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

         price_range     busiest_day   busiest_hours  max_order  \
0  Mid-range product  Regularly Busy     Most Orders         32   
1  Mid-range product  Regularly Busy  Average Orders         32   
2  Mid-range product     Busiest Day  Average Orders          5   
3  Mid-range product  Regularly Busy     Most Orders          3   
4  Mid-range product      Least Busy  Average Orders          3   

       loyalty_flag  
0  Regular Customer  
1  Regular Customer  
2      New Customer  
3      New Customer  
4      New Customer

## Q. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [13]:
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
Loyal Customer    10.386336
New Customer      13.294670
Regular Customer  12.495717

### A. From the above result, we can infer that on average New Customers spend the most and Loyal Customers spend the least.

## Q. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
## If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
## If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [15]:
# Finding the mean for all users and creating a column to display the mean
df_ords_prods_merge['average_spent'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/z8/12vdbzn57fx_ky9_g3vm3lv80000gn/T/ipykernel_25838/3808926498.py:2: FutureWarning: The provided callable <function mean at 0x107542200> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['average_spent'] = df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [17]:
# checking column creation
df_ords_prods_merge[['prices', 'average_spent']].head(10)

prices  average_spent
0     5.8       6.935811
1     5.8       6.935811
2     5.8       7.930208
3     5.8       4.972414
4     5.8       4.972414
5     5.8       6.935398
6     5.8       5.957576
7     5.8       6.680000
8     5.8       7.162500
9     5.8       7.162500

In [18]:
# creating a new column 'spending_flag' by using the results of column 'average_spent'
df_ords_prods_merge.loc[df_ords_prods_merge['average_spent'] < 10, 'spending_flag'] = 'Low Spender'
df_ords_prods_merge.loc[df_ords_prods_merge['average_spent'] >= 10, 'spending_flag'] = 'High Spender'

In [19]:
# checking column creation
df_ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_the_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range', 'busiest_day', 'busiest_hours',
       'max_order', 'loyalty_flag', 'average_spent', 'spending_flag'],
      dtype='object')

In [22]:
df_ords_prods_merge[['user_id', 'average_spent', 'spending_flag']].head(15)

user_id  average_spent spending_flag
0       138       6.935811   Low Spender
1       138       6.935811   Low Spender
2       709       7.930208   Low Spender
3       764       4.972414   Low Spender
4       764       4.972414   Low Spender
5       777       6.935398   Low Spender
6       825       5.957576   Low Spender
7       910       6.680000   Low Spender
8      1052       7.162500   Low Spender
9      1052       7.162500   Low Spender
10     1379       5.813636   Low Spender
11     1480       5.081633   Low Spender
12     1480       5.081633   Low Spender
13     1480       5.081633   Low Spender
14     1494       6.803797   Low Spender

In [21]:
# checking the frequency for High and Low Spenders
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low Spender     31770614
High Spender      634245
Name: count, dtype: int64

## Q. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
## If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
## If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
## If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [23]:
# creating a new column that records the median of 'days_since_prior_order'
df_ords_prods_merge['order_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/z8/12vdbzn57fx_ky9_g3vm3lv80000gn/T/ipykernel_25838/3743713989.py:2: FutureWarning: The provided callable <function median at 0x107677c40> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['order_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [25]:
# checking column creation
df_ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_the_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range', 'busiest_day', 'busiest_hours',
       'max_order', 'loyalty_flag', 'average_spent', 'spending_flag',
       'order_frequency'],
      dtype='object')

In [26]:
df_ords_prods_merge[['order_frequency']].head(10)

order_frequency
0              8.0
1              8.0
2              8.0
3              9.0
4              9.0
5             11.0
6             20.0
7              6.0
8             10.0
9             10.0

In [28]:
# creating a new column 'order_frequency_flag' by using the values from 'order_frequency'
df_ords_prods_merge.loc[df_ords_prods_merge['order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent Customer'
df_ords_prods_merge.loc[(df_ords_prods_merge['order_frequency'] > 10) & (df_ords_prods_merge['order_frequency'] <= 20), 'order_frequency_flag'] = 'Regular Customer'
df_ords_prods_merge.loc[df_ords_prods_merge['order_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent Customer'

In [29]:
# checking column creation
df_ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_the_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range', 'busiest_day', 'busiest_hours',
       'max_order', 'loyalty_flag', 'average_spent', 'spending_flag',
       'order_frequency', 'order_frequency_flag'],
      dtype='object')

In [31]:
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent Customer        21559853
Regular Customer          7208564
Non-frequent Customer     3636437
NaN                             5
Name: count, dtype: int64

In [32]:
df_ords_prods_merge[['user_id', 'loyalty_flag', 'spending_flag', 'order_frequency_flag']].head(20)

user_id      loyalty_flag spending_flag order_frequency_flag
0       138  Regular Customer   Low Spender    Frequent Customer
1       138  Regular Customer   Low Spender    Frequent Customer
2       709      New Customer   Low Spender    Frequent Customer
3       764      New Customer   Low Spender    Frequent Customer
4       764      New Customer   Low Spender    Frequent Customer
5       777  Regular Customer   Low Spender     Regular Customer
6       825      New Customer   Low Spender     Regular Customer
7       910  Regular Customer   Low Spender    Frequent Customer
8      1052  Regular Customer   Low Spender    Frequent Customer
9      1052  Regular Customer   Low Spender    Frequent Customer
10     1379      New Customer   Low Spender     Regular Customer
11     1480    Loyal Customer   Low Spender    Frequent Customer
12     1480    Loyal Customer   Low Spender    Frequent Customer
13     1480    Loyal Customer   Low Spender    Frequent Customer
14     1494  Regular Customer   Low Spender    Frequent Customer
15     1494  Regular Customer   Low Spender    Frequent Customer
16     1494  Regular Customer   Low Spender    Frequent Customer
17     1540  Regular Customer   Low Spender    Frequent Customer
18     1540  Regular Customer   Low Spender    Frequent Customer
19     1540  Regular Customer   Low Spender    Frequent Customer

# Export Data

In [33]:
df_ords_prods_merge.to_pickle(os.path.join(path, '2 Data', 'Prepared Data', 'ords_prods_merge_group.pkl'))